src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
+        constexpr int num_permutations = 1 << sizeof(key_type); // 4 bytes gives up to 16 search keys
+        const key_type upper_key = StringIndex::create_key(upper_value, string_offset);
+        const key_type lower_key = StringIndex::create_key(lower_value, string_offset);
+        for (int p = 0; p < num_permutations; ++p) {
+            // this might still be incorrect due to unicode characters crossing the 4 byte key size
+            const key_type key = generate_key(upper_key, lower_key, p);
+            const bool new_key = keys_seen.insert(key).second;
```

Comment:
That it would be a "real" performance regression? No longer sure what you prefer? Between optimizing the choice of containers and disabling the index entirely is a pretty big gap.